[Feature] Add TabbedCommandBar (ribbon) control

yoshiask opened the pull request:
<!-- 🚨 Please Do Not skip any instructions and information mentioned below as they are all required and essential to evaluate and test the PR. By fulfilling all the required information you will be able to reduce the volume of questions and most likely help merge the PR faster 🚨 -->

## Fixes #3259
<!-- Add the relevant issue number after the "#" mentioned above (for ex: Fixes #1234) which will automatically close the issue once the PR is merged. -->


<!-- Add a brief overview here of the feature/bug & fix. -->

## PR Type
What kind of change does this PR introduce?
<!-- Please uncomment one or more that apply to this PR. -->

<!-- - Bugfix -->
Feature
<!-- - Code style update (formatting) -->
<!-- - Refactoring (no functional changes, no api changes) -->
<!-- - Build or CI related changes -->
<!-- - Documentation content changes -->
Sample app changes
<!-- - Other... Please describe: -->


## What is the current behavior?
There is no ribbon or ribbon-like control for UWP. The current solution is to put CommandBars inside of a Pivot.

## What is the new behavior?
Adds a TabbedCommandBar, which internally is almost identical to the current solutions, but provides a much friendlier developer experience.


## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tested code with current [supported SDKs](../readme.md#supported)
- [ ] Pull Request has been submitted to the documentation repository [instructions](..\contributing.md#docs). Link: https://github.com/MicrosoftDocs/WindowsCommunityToolkitDocs/pull/405
- [ ] Sample in sample app has been added / updated (for bug fixes / features)
    - [x] Icon has been created (if new sample) following the [Thumbnail Style Guide and templates](https://github.com/windows-toolkit/WindowsCommunityToolkit-design-assets)
- [ ] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Header has been added to all new source files (run *build/UpdateHeaders.bat*)
- [x] Contains **NO** breaking changes

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. 
     Please note that breaking changes are likely to be rejected within minor release cycles or held until major versions. -->


## Other information
I forgot to do my work on a feature branch, so this PR is basically a duplicate of #3551, with a little bit of cleanup.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/TabbedCommandBar/TabbedCommandBar.xaml:
```diff
@@ -0,0 +1,290 @@
+        <Setter Property="PaneToggleButtonStyle" Value="{StaticResource PaneToggleButtonStyle}" />
+        <Setter Property="IsTabStop" Value="False" />
+        <Setter Property="CompactPaneLength" Value="{ThemeResource NavigationViewCompactPaneLength}" />
+        <Setter Property="Background" Value="{ThemeResource SystemControlChromeMediumLowAcrylicWindowMediumBrush}" />
```
I know this if fancy for apps which want to extend this into the titlebar, but we should default to the standard use-case of it fitting in with dropping into a blank app and use the existing resources like TabView or NavigationView use currently.

Developers can override this if they need to.

yoshiask replied:
The latest commit changes the default style to be the chrome (non-acrylic) equivalent of the previous defaults. That should solve this (and also not look as weird on the sample page). I supplied the acrylic version as a style that's based on the default style.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/TabbedCommandBar/TabbedCommandBar.xaml:
```diff
@@ -0,0 +1,290 @@
+                                                   Normal="{StaticResource NormalTabTemplate}" />
+
+    <StaticResource x:Key="TabbedCommandBarContentBackground" ResourceKey="SystemControlChromeLowAcrylicWindowBrush" />
+    <Style TargetType="controls:TabbedCommandBar">
```
Please follow the suggested guidance we use elsewhere of naming this and setting the default separately (and also splitting out the control template). See examples in things like the TokenizingTextBox.

yoshiask replied:
I think this is fixed in the latest commit. I'll double-cehc

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/TabbedCommandBar/TabbedCommandBarItemTemplateSelector.cs:
```diff
@@ -0,0 +1,37 @@
+    /// <summary>
+    /// <see cref="DataTemplateSelector"/> used by <see cref="TabbedCommandBar"/> for determining the style of normal vs. contextual <see cref="TabbedCommandBarItem"/>s.
+    /// </summary>
+    public class TabbedCommandBarItemTemplateSelector : DataTemplateSelector
```
Did you investigate the style selector vs. data template?

yoshiask replied:
I've been trying to figure out how to override those NavigationViewItem resources in a style. I think it has to be a template for it to work

michael-hawker replied:
Ah, I see what you mean around the overridden resources... good point. Hmm. I'll think about that, for now we can leave it.

michael-hawker reviewed (commented):
Quick comments

yoshiask commented:
I'm using the package generated by the pipeline and I can see a few bugs that I'm not sure how to address.
- The animation when switching tabs is just gone. The TabbedCommandBarItems just appear in place
- The styles that I defined for the controls (e.g. `DefaultTabbedCommandBarItemStyle`, `AcrylicTabbedCommandBar`) aren't available in projects that just reference the Toolkit package. Attempting to use one of those styles in an app results in a "resource not found" exception.

yoshiask commented:
@michael-hawker Just pinging you since it's been a while, did you see my last comment?
https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3556#issuecomment-726656619

michael-hawker commented:
So sorry @yoshiask, I missed your comment after .NET Conf.

For your issues with the Sample App and stand-alone. The Sample App uses WinUI 2, so it's including their style overrides compared to when just run with the Toolkit and the system styles. This is something we have to address in the future better (opened #3598 to track this). It might be that we have to check for what this conflict/difference may be?

Do those two styles use specific resources found in the WinUI 2 styles?

yoshiask commented:
No, all styles use system resources. Also just want to reiterate that the issue isn't coming from inside the custom styles. The issue is that the styles I wrote don't seem to be accessible outside of the Toolkit resource dictionary

yoshiask commented:
For example, the DefaultTabbedCommandBarItemStyle _is_ the style that couldn't be found.

michael-hawker commented:
@yoshiask I see, you're saying for accessing in the code a developer would be using in their app, eh?

Does it work for any other style for any other controls in the Toolkit? This may be a general issue not specific to your component. This also may be related to how we don't require inclusion of resources in the app like WinUI does as well.

I know in more explicit cases where we wanted alternate styles for folks to leverage, we just had them as loose XAML files which we then had folks include explicitly like with the alternate styles for `InAppNotification`:

https://github.com/windows-toolkit/WindowsCommunityToolkit/blob/2a4938d9f11faf6ff3630294470194598ae4da3b/Microsoft.Toolkit.Uwp.SampleApp/SamplePages/InAppNotification/InAppNotificationXaml.bind#L14-L17

michael-hawker commented:
Going to close and re-open to try and re-trigger CI?

yoshiask commented:
I merged master into the branch, so there aren't any merge conflicts now. I think this will fix the CI.

yoshiask commented:
The feature is not quite done, but I think it's close enough to start reviewing and requesting changes.

michael-hawker commented:
@yoshiask I just noticed that @dpaulino's Nightingale has a similar control! 🙂

![image](https://user-images.githubusercontent.com/24302614/104388269-b8d6cf80-54ed-11eb-876e-d410b9fc4734.png)

mdtauk commented:
Comparing both @yoshiask and @dpaulino Ribbon controls - are there big feature gaps, or does one feature a better implementation and API than the other?

yoshiask commented:
> @yoshiask I just noticed that @dpaulino's Nightingale has a similar control! 🙂
> 
> ![image](https://user-images.githubusercontent.com/24302614/104388269-b8d6cf80-54ed-11eb-876e-d410b9fc4734.png)
> 

I can't say much about the internal implementation in Nightingale, but I do know that it's a Pivot with command bars. It's basically the same as this PR, with the major difference being the use of a NavigationView instead of a Pivot. I designed the API surface for the WCT control to be similar to how it would be in a custom solution like Nightingale's,  but with sensible naming and additional style changes (like right aligning the overflow button and overriding styles for common but non-AppBar controls).

yoshiask commented:
tl:dr; This TabbedCommandBar control was designed to be a mostly drop-in replacement for existing solutions like Nightingale's, but with the benefits of custom ribbon styles and ribbon-specific features (like tab/command bar footers and eventually Quick Access and collapsing the ribbon)

dpaulino commented:
Is this a bug?
![image](https://user-images.githubusercontent.com/7649589/104537565-0c6d1a00-55cf-11eb-8211-9a2f3ad56cc3.png)

yoshiask commented:
Uh... yeah. How did you manage that? Maybe it's a bug in WinUI?

Kyaa-dost commented:
@yoshiask could you please elaborate on why is this a bug?

yoshiask commented:
> @yoshiask could you please elaborate on why is this a bug?

Somehow two NavigationViewItems appear to be selected at the same time

dpaulino commented on Microsoft.Toolkit.Uwp.UI.Controls.Design/TabbedCommandBarMetadata.cs:
```diff
@@ -0,0 +1,36 @@
+            AddCallback(typeof(TabbedCommandBar),
+                b =>
+                {
+                    //b.AddCustomAttributes(nameof(TabbedCommandBar.SomePropertyHere),
+                    //    new CategoryAttribute(Properties.Resources.CategoryCommon)
+                    //    );
```
Is this still needed? if not we should delete

yoshiask replied:
I'm actually not sure what this class is for. Do you have any insights?

dpaulino replied:
Oh, I'm not sure. Did you add this? or something else added this file?

yoshiask replied:
Every control in the Toolkit appears to have a metadata class like this. Maybe it has something to do with the 'Properties' tab in the sample app?

michael-hawker replied:
The Metadata class is to help organize properties within the Visual Studio Properties panel for the Designer. You should follow the pattern of other controls here and split out any custom properties into the proper categories you'd think they'd fit within the VS Designer.

yoshiask commented on Microsoft.Toolkit.Uwp.UI.Controls/TabbedCommandBar/TabbedCommandBar.cs:
```diff
@@ -0,0 +1,83 @@
+            _ribbonContent = GetTemplateChild("PART_RibbonContent") as ContentControl;
+            _tabChangedStoryboard = GetTemplateChild("TabChangedStoryboard") as Storyboard;
+
+            SelectedItem = MenuItems.FirstOrDefault();
```
My understanding is that the NavigationView base handles MenuItems and initializes it with an empty collection in its constructor.

Kyaa-dost commented:
> > @yoshiask could you please elaborate on why is this a bug?
> 
> Somehow two NavigationViewItems appear to be selected at the same time

That's interesting because I am not able to repro this bug. 

![image](https://user-images.githubusercontent.com/35208324/104661571-a68b9b80-567d-11eb-9852-708537273529.png)

dpaulino commented:
@Kyaa-dost here is the repro: 

https://user-images.githubusercontent.com/7649589/104671552-e4df8580-5692-11eb-834a-45ec13f131d6.mp4

dpaulino commented:
Also, notice in the video I uploaded that the content flashes when resizing the grid. Is that expected? Or is that a bug?

Kyaa-dost commented:
@dpaulino I am able to repro now and thanks for the screen record. 

Also, correct it does seem like a bug to me and I have opened an issue #3674

dpaulino commented:
> @dpaulino I am able to repro now and thanks for the screen record. 
> 
> 
> 
> Also, correct it does seem like a bug to me and I have opened an issue #3674 

@Kyaa-dost Is the bug specifically with this tabbed command bar or with the entire toolkit? I think I've only seen the flashing bug in this PR's tabbed command bar.

Kyaa-dost commented:
@dpaulino seems consistent to me. Check out the video I posted in #3674

yoshiask commented:
@michael-hawker I just saw this article, https://www.reflectionit.nl/blog/2021/xaml-lightweight-styling-done-right, is this something we could be using here? I think it might be helpful in the creation of AppBar* controls for existing controls

juliusfriedman commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/TabbedCommandBar/TabbedCommandBar.bind:
```diff
@@ -19,7 +19,7 @@
                     <AppBarSeparator />
                     <AppBarElementContainer>
-                        <ComboBox SelectedIndex="0">
+                        <ComboBox SelectedIndex="0" Width="175">
```
eww

yoshiask replied:
Yeah, I don't like it either, but without it the ComboBox changes width depending on how long the selected text is.

dpaulino replied:
This is just a sample page right? So the hardcoded value seems fine

mdtauk replied:
What about a MinWidth, or a MaxWidth?

yoshiask replied:
I'll make it use `MinWidth` instead. `MaxWidth` isn't what I'm going for, since it would still resize if the text is shorter than the longest string in the menu.

michael-hawker commented:
> @michael-hawker I just saw this article, https://www.reflectionit.nl/blog/2021/xaml-lightweight-styling-done-right, is this something we could be using here? I think it might be helpful in the creation of AppBar* controls for existing controls

Ideally we'd bring in @rudyhuyn's XamlPlus to the toolkit yes, but don't think that'll happen for 7.0?

You're thinking for providing better customization of the controls to sit within the toolbar, eh?

michael-hawker commented:
@yoshiask have you had a chance to look at the bug @dpaulino found and resolve his comments?

yoshiask commented:
@michael-hawker No, I haven't been able to work on any VS projects at all because of some critical bugs that I can't seem to resolve. I've just reinstalled Windows and I'm currently going through the process of reinstalling VS.

Kyaa-dost commented:
@yoshiask no worries, just let us know of any updates and we should be good to merge this PR soon.

michael-hawker commented:
I was able to reproduce the bug but it took a little while. @dpaulino do you know if there was a specific set of steps you took to reproduce it every time consistently? And were you in Debug or Release mode ([as you never know what to expect](https://github.com/microsoft/microsoft-ui-xaml/issues/2521))?

I was trying to see if I could reproduce on a NavigationView in XAML Studio, but it could be an older version as well...

```xml
<Page
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    xmlns:muxc="using:Microsoft.UI.Xaml.Controls"
    mc:Ignorable="d">

    <muxc:NavigationView PaneDisplayMode="Top">
        <muxc:NavigationView.MenuItems>
            <muxc:NavigationViewItem Content="Home"/>
            <muxc:NavigationViewItem Content="Insert"/>
        </muxc:NavigationView.MenuItems>
    </muxc:NavigationView>
</Page>
```

In either case, I don't know how common it'll be yet; so I'm happy if we can forward-fix that issue later if it turns out to be in the Toolkit. 

@yoshiask anything else you wanted to wrap up on this? Looks/API wise it's looking great from my perspective, so I'm just going to look through the code one last time, though I know @dpaulino took a look already which was great. 🎉

yoshiask commented:
There are a few things I'd like to add, mostly styles for SplitButton and DropDownButton. I think I can borrow from styles from Windows Calculator, since the "Trigonometry" button in Scientific mode is pretty much exactly how I think a drop-down button should look in a ribbon. Other than that, the API surface is pretty much final.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/TabbedCommandBar/TabbedCommandBar.xaml:
```diff
@@ -0,0 +1,299 @@
+    <DataTemplate x:Key="ContextualTabTemplate">
+        <NavigationViewItem Content="{Binding Header}">
+            <NavigationViewItem.Resources>
+                <SolidColorBrush x:Key="TopNavigationViewItemForeground" Color="{ThemeResource SystemAccentColor}" />
```
Is this just adjusting the text? Looking at office looks like the new pattern is to just color the text:

Selected
![image](https://user-images.githubusercontent.com/24302614/106988122-cb76aa00-6723-11eb-97f6-a2831514815d.png)

vs Unselected
![image](https://user-images.githubusercontent.com/24302614/106988138-d598a880-6723-11eb-8479-1ad96edd1949.png)

And Office Online adds a bit of a different background shade to them:
![image](https://user-images.githubusercontent.com/24302614/106988229-1690bd00-6724-11eb-93d5-5f4fc374b319.png)

(Which is probably what we may want to do too.)

I like the idea of using the AccentColor as the default, but we should expose these as named resources within our dictionary set to these values so they're more easily overridable. [TabView had examples of this](https://github.com/windows-toolkit/WindowsCommunityToolkit/blob/rel/6.1.1/Microsoft.Toolkit.Uwp.UI.Controls/TabView/TabView.xaml).

yoshiask replied:
![image](https://user-images.githubusercontent.com/7549325/106990538-0e437c80-673a-11eb-9a47-c5a1ce71dbbc.png)
Yes, right now it only adjusts the text foreground. I'll set up more default resources and look into changing the hover/selected background colors.

michael-hawker replied:
Thanks Yoshi, it's looking really slick. I think the main thing would be if we just have a slightly lighter tone for the background of the contextual item, you can see above it's like that in the web world even if it's not selected (it's the same if it's selected to). This is a hover example:

![image](https://user-images.githubusercontent.com/24302614/107003280-5e720d00-6741-11eb-9fad-e2cdc121aacc.png)

And if we name them as resources, folks can override them if they'd like.

I think the online versions are more akin to this pattern and more up-to-date with the direction of styling, so it'd be cool to mimic that a bit more with having the toned background to make the contextual tab stand-out a bit more beyond the accent color.

This is a real nitpicky detail, so having the Visibility binding working is the main priority (see my updated sample example I pasted in the other comment). If we could have that toggleswitch just have the Contextual tab appear with the binding all in XAML, I think we're golden. 🙂

mdtauk replied:
The Office ribbon uses Semibold for selected tabs, which may be important to include, if colour is used for contextual tabs, which wont work so well with a lower opacity for unselected tab headers.

yoshiask replied:
> The Office ribbon uses Semibold for selected tabs, which may be important to include, if colour is used for contextual tabs, which wont work so well with a lower opacity for unselected tab headers.

I did consider adding that, but I couldn't figure out how. At the moment, tab headers aren't required to be text, so I can't think of a clean way to make text bold without reaching into the visual tree using code behind. I'm sure it's possible, but I'm not experienced enough with XAML to do it.

yoshiask replied:
> Thanks Yoshi, it's looking really slick. I think the main thing would be if we just have a slightly lighter tone for the background of the contextual item, you can see above it's like that in the web world even if it's not selected (it's the same if it's selected to). This is a hover example:
> 
> ![image](https://user-images.githubusercontent.com/24302614/107003280-5e720d00-6741-11eb-9fad-e2cdc121aacc.png)
> 
> And if we name them as resources, folks can override them if they'd like.

Here's what I've done so far. Unfortunately, Office Online doesn't have an official dark mode, so I've just picked the UWP brush resource that matched in light mode and looked plausible in dark mode.
![image](https://user-images.githubusercontent.com/7549325/107163356-b93c7c00-696e-11eb-9610-94eec218f2eb.png)
![image](https://user-images.githubusercontent.com/7549325/107163377-c6596b00-696e-11eb-9ccb-492d644dea6c.png)

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/TabbedCommandBar/TabbedCommandBarItem.cs:
```diff
@@ -0,0 +1,118 @@
+        /// </summary>
+        public static readonly DependencyProperty HeaderProperty = DependencyProperty.Register(
+            nameof(Header),
+            typeof(string),
```
Since `NavigationViewItem.Content` is an object, our `Header` here should be an `object` as well. That way if someone wants to customize it and add an icon or something next to their text they can pass that in XAML with a different control.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/TabbedCommandBar/TabbedCommandBarItem.cs:
```diff
@@ -0,0 +1,118 @@
+            nameof(Footer),
+            typeof(UIElement),
+            typeof(TabbedCommandBarItem),
+            new PropertyMetadata(new Border()));
```
Reason for defaulting to a Border here? I was trying to see where this was referenced in the XAML?

yoshiask replied:
I just didn't want it to default to `null`, is that what it should do instead?

michael-hawker replied:
I think as long as there's not visual glitches or anything, the Footer being null by default is fine. [As an example the default value for ListViewBase.Footer is null](https://docs.microsoft.com/uwp/api/windows.ui.xaml.controls.listviewbase.footer). [NavigationView.PaneFooter is also null](https://docs.microsoft.com/uwp/api/windows.ui.xaml.controls.navigationview.PaneFooter).

yoshiask replied:
Setting it to null didn't cause any issues. However, I removed the Footer property for now since it's not implemented. Adding it would require a retemplate of the CommandBar, which I really want to avoid at least until https://github.com/microsoft/microsoft-ui-xaml-specs/pull/108 is finalized.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/TabbedCommandBar/TabbedCommandBarItem.cs:
```diff
@@ -0,0 +1,118 @@
+            _primaryItemsControl = GetTemplateChild("PrimaryItemsControl") as ItemsControl;
+            if (_primaryItemsControl != null)
+            {
+                _primaryItemsControl.HorizontalAlignment = HorizontalAlignment.Stretch;
```
It'd probably be better to setup a Binding here so that the binding flows through, but I just learned how to do this and haven't added my hacky helper yet... 🤔 May need to make a follow-up note later for myself. 🙂

yoshiask replied:
I can't remember exactly why, but there was a reason why I hardcoded it. It's not a value that should be modified by the user.

yoshiask replied:
I remember why. `_primaryItemsControl.HorizontalAlignment` determines the alignment of the commands in the CommandBar. Now that I'm thinking about it, that would be a useful thing to expose (particularly for localization in the future). I've created the `CommandAlignment` property on `TabbedCommandBar`.

michael-hawker reviewed (changes requested):
Hey @yoshiask did a pass at the code, looking really solid and straight-forward for the main components. Really nice to see it all come together. Few (hopefully) minor comments. I think my biggest question was just knowing if the Contextual Tabs are working (and if so added that to the sample)? Do you have a screenshot of what they look like now?

Is your machine back up and running?

michael-hawker commented:
@yoshiask just saw your new comment. We can always add the additional styles as a follow-on PR if you'd like to snap this in. It'd be great to get the base of it in before our final preview middle of next week (🤞).

Note though too we're moving controls around, so there's also going to be a point where you'll need to move to the new location, thinking our dev branch may merge on Monday or Tuesday? If we can get this PR wrapped up by then, we can do the initial move for you from the current main before that comes in.

yoshiask commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/TabbedCommandBar/TabbedCommandBarPage.xaml.cs:
```diff
@@ -0,0 +1,34 @@
+    /// <summary>
+    /// An empty page that can be used on its own or navigated to within a Frame.
+    /// </summary>
+    public sealed partial class TabbedCommandBarPage : Page
```
At the moment, setting `IsContextual` to `True` on a tab just applies the special style to it. In the future, I want to at least add ContextualTabGroups and force all tabs with `IsContextual` to appear at the end of the list.

michael-hawker replied:
Thanks @yoshiask, I was being dumb earlier and forgot I could try adding it to the live editor in the sample. 🤦‍♂️

I tried adding this snippet to the end, but the `TabbedCommandBarItem` seems to ignore it's `Visibility` property. It's probably because it's not bound to the underlying NavigationViewItem?

```xml
      xmlns:converters="using:Microsoft.Toolkit.Uwp.UI.Converters"
      mc:Ignorable="d">
      <Page.Resources>
        <converters:BoolToVisibilityConverter x:Key="BoolToVisibilityConverter"/>
      </Page.Resources>

                <controls:TabbedCommandBarItem Header="Contextual" 
                                               IsContextual="True"
                                               Visibility="{Binding ElementName=ContextualToggle, Path=IsOn, Converter={StaticResource BoolToVisibilityConverter}}">
                </controls:TabbedCommandBarItem>
            </controls:TabbedCommandBar.MenuItems>
        </controls:TabbedCommandBar>
        <ToggleSwitch 
            x:Name="ContextualToggle"
            OffContent="Contextual Tab Off" OnContent="Contextual Tab On"/>
```

I also noticed `IsContextual` doesn't work bound either, so it may be worth just making it a regular property if we don't expect it to support being updated dynamically (which we wouldn't).

michael-hawker replied:
Oh, you may not need the converter here, I thought I needed it at first, but was binding to the wrong property.

yoshiask replied:
Since the TabbedCommandBar handles the navigation/selection, how would we do this? The TabbedCommandBar has no way on knowing when the currently selected TabbedCommandBarItem becomes collapsed.

michael-hawker replied:
@yoshiask I would imagine the underlying NavigationView would have some logic for this hopefully, but we'd have to test it out by making sure this example of binding to the visibility works first by actually changing the visibility of the underlying NavigationViewItem? Otherwise, in TabbedCommandBarItem itself you can [register to listen for a VisibilityChanged](https://docs.microsoft.com/uwp/api/windows.ui.xaml.dependencyobject.registerpropertychangedcallback?view=winrt-19041#Windows_UI_Xaml_DependencyObject_RegisterPropertyChangedCallback_Windows_UI_Xaml_DependencyProperty_Windows_UI_Xaml_DependencyPropertyChangedCallback_) event and do something based on that.

yoshiask replied:
I added a ToggleSwitch to the sample page, and it does work, but there's some weird initial behavior. The switch won't work until the tab bound to it is selected at least once.

https://user-images.githubusercontent.com/7549325/107172442-fdd61080-698a-11eb-95b3-e990ab1afeac.mp4

yoshiask commented on Microsoft.Toolkit.Uwp.UI.Controls/Microsoft.Toolkit.Uwp.UI.Controls.csproj:
```diff
@@ -44,6 +44,11 @@
     <LangVersion>8.0</LangVersion>
   </PropertyGroup>
 
+  <ItemGroup>
+    <None Remove="TabbedCommandBar\TabbedCommandBar.xaml" />
+    <None Remove="TabbedCommandBar\TabbedCommandBarItem.xaml" />
+  </ItemGroup>
+
```
That's odd. I'm not even sure what that's doing or how it got there.

michael-hawker replied:
Probably happened through some random VS operation, they should be able to be safely removed and still work fine.

yoshiask commented on Microsoft.Toolkit.Uwp.UI.Controls/TabbedCommandBar/TabbedCommandBar.cs:
```diff
@@ -0,0 +1,83 @@
+            {
+                // This code is a hack, but it's necessary if binding doesn't work.
+                // RibbonContent might be null here, there should be a check
+                _ribbonContent.Content = MenuItems[System.Math.Min(MenuItems.Count - 1, MenuItems.IndexOf(navItem))];
```
Yeah, probably. (Note to self: use `// TODO` more often)

yoshiask replied:
Turns out this code isn't necessary anymore, since I can bind directly to the underlying NavigationView's `SelectedItem` like so:
```
<ContentControl x:Name="PART_RibbonContent" Content="{TemplateBinding SelectedItem}"/>
```

yoshiask commented:
My machine is working now, I'm just trying to find time to squeeze work in.

Contextual tabs have the absolute bare minimum functionality, since you can bind to visibility to hide it and there's an `IsContextual` property that [at the moment] applies the special style. I put a screenshot in a reply to your review comment.

I might have time over the weekend to finally finish up the PR so it can be merged by Monday.

yoshiask commented on Microsoft.Toolkit.Uwp.UI.Controls/TabbedCommandBar/TabbedCommandBar.cs:
```diff
@@ -0,0 +1,83 @@
+    [TemplatePart(Name = "PART_RibbonNavigationView", Type = typeof(NavigationView))]
+    [TemplatePart(Name = "PART_RibbonContent", Type = typeof(ContentControl))]
+    [TemplatePart(Name = "PART_TabChangedStoryboard", Type = typeof(Storyboard))]
+    public class TabbedCommandBar : NavigationView
```
Are we adding WinUI for this release? My only concern would be whether we inherit from the system NavigationView or the MUXC one.

michael-hawker replied:
We're going to be having it all-up in the new uber-Controls meta-package, so it's safe to switch to the MUXC one, though I'm not sure if we're going to target 2.4 or 2.5 yet, that's the next thing for me to find out. It'll be safe to continue what you're doing building on WUX; but not sure if you've been trying both and which was easier for you, so wanted to give you a heads up. I'll try and get an answer tomorrow on which version we're going to target.

yoshiask replied:
Now that I've done more debugging and stress testing, it appears that the system NavigationView has quite a few bugs. I don't know if any of them are fixed in the WinUI version, but at least then there's hope of it being addressed.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/TabbedCommandBar/TabbedCommandBar.xaml:
```diff
@@ -0,0 +1,297 @@
+        <NavigationViewItem Content="{Binding Header}" />
+    </DataTemplate>
+    <DataTemplate x:Key="ContextualTabTemplate">
+        <NavigationViewItem Content="{Binding Header}">
```
I'm thinking this is where the Visibility issue is, as the the CommandBarItem's visibility isn't tied to what's being displayed in the NavigationView.

I tested in XAML Studio that both the WUX & MUXC NavigationViews at least will respect a NavigationViewItem's Visibility property and hide it in the Top level navigation mode and that it'll respond to binding. You may get away with just a `Visibility="{Binding Visibility}"` here, but I'm not sure? Would there be other things like `IsEnabled` that need to be mapped too?

Only case I'm not sure what'll happen if the Contextual Tab is selected and then disappears... But maybe your selection changed event logic handles that?

yoshiask replied:
> Only case I'm not sure what'll happen if the Contextual Tab is selected and then disappears... But maybe your selection changed event logic handles that?

Oops, I hadn't thought about that when I switched to binding the ribbon content control directly to the SelectedItem. I think the old SelectionChanged logic could have been modified to handle it pretty easily (it just needs to check if the item's visibility is `Collapsed` and instead select the first ribbon item) @michael-hawker Do I need to revert back to using SelectionChanged, or is there a way to add that logic and still use a binding?

michael-hawker replied:
I would try binding the Visibility here like I mentioned on line 13 first to see if that works? It may just allow the passthru to work?

yoshiask replied:
Yup, that works. It also can handle when the currently selected tab is set to hidden.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/TabbedCommandBar/TabbedCommandBar.xaml:
```diff
@@ -267,25 +266,24 @@
-                        </ContentControl>
+
+                        <Border x:Name="PART_RibbonContentBorder"
+                                Background="{TemplateBinding ItemBackground}" Grid.Row="1"
```
```cs
SelectionChanged += (NavigationView sender, NavigationViewSelectionChangedEventArgs args) =>
            {
                _ribbonContentBorder.Background = (sender.SelectedItem as Control).Background;
```
I'm a bit confused, isn't this the thing you're setting above in the SelectionChanged event? So then what's the point of using the binding and exposing the ItemBackground property?

yoshiask replied:
Thanks for pointing that out, I forgot to remove the binding from when I was debugging it.

michael-hawker commented:
Thanks @yoshiask, I'm off for the night, but poked a bit at the sample we could have for Contextual Tabs, hope that helps point things in the right direction for finalizing that scenario!

michael-hawker commented:
@yoshiask @dpaulino found solid repro steps for the bug:

1) Load Sample
2) Click on Insert
3) Then smoosh area with GridSplitter so 'More' appears
4) Click on More and click 'Home'
5) Expand area w/ GridSplitter
6) Voila! Both Selected

I wasn't able to reproduce using the same steps in XAML Studio with the built-in NavigationView or the MUXC one, so seems specific to something going on w/ usage in TabbedCommandBar?

yoshiask commented on Microsoft.Toolkit.Uwp.UI.Controls/TabbedCommandBar/TabbedCommandBar.cs:
```diff
@@ -0,0 +1,83 @@
+        /// <summary>
+        /// Gets or sets the brush to use as the background for content
+        /// </summary>
+        public Brush ContentBackground
```
Renamed to `ItemBackground`, but that's still not great. Maybe `RibbonBackground` would be better?

michael-hawker replied:
We'd want to avoid using the term `Ribbon` anywhere, so `CommandBarBackground` would probably be better? Is it supposed to be the top bar or bottom bar's background color? I think that's what I was trying to understand.

yoshiask replied:
It's the background behind the CommandBar. I decided to just remove it and always have it set to the selected item's background. Maybe in the future it could be put back, but I just don't see a use for it (since it only appears briefly during the transition animation between tabs).

yoshiask commented:
Okay, I'm officially calling it here. I'm not adding any new stuff (as much as I wanted to make the ribbon collapsible). I believe I've addressed all the feedback, but feel free to unresolve a thread if there's more work to be done there.

Kyaa-dost commented:
@yoshiask Thanks for all the work ❤️

 Will have to look more into this as we are still able to repro this error highlighted by the community. 

![image](https://user-images.githubusercontent.com/35208324/107287758-3de2d500-6a17-11eb-926e-ebd716b66336.png)

michael-hawker commented:
Thanks @Kyaa-dost, yeah I'm not overly worried about that bug as it's a bit tricky to actually reproduce unless you follow an explicit set of steps. I think it's something we can address later or even in 7.1 if we find out it's a bigger issue.

I plan to give this another whirl and double-check the contextual tab updates tonight. So excited to get this in for the release! 🎉

Great work @yoshiask!

michael-hawker commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/TabbedCommandBar/TabbedCommandBar.bind:
```diff
@@ -78,7 +93,25 @@
                     </controls:TabbedCommandBarItem.SecondaryCommands>
                 </controls:TabbedCommandBarItem>
+                <controls:TabbedCommandBarItem Header="Shape Format" IsContextual="True"
+                                               Visibility="{Binding ElementName=ContextualToggle, Path=IsOn}">
```
Alright, inverting the binding paradigm works (albeit a bit oddly)... We probably need to go with this for now, and just document that it should be toggled via code-behind or this method? (or behaviors may work too?)

Thoughts?

```xml
<Page xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
      xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
      xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
      xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
      xmlns:controls="using:Microsoft.Toolkit.Uwp.UI.Controls"
      xmlns:converters="using:Microsoft.Toolkit.Uwp.UI.Converters"
      mc:Ignorable="d">
      <Page.Resources>
        <converters:VisibilityToBoolConverter x:Key="BoolVisConverter"/>
      </Page.Resources>
    <Grid>
        <Grid.RowDefinitions>
            <RowDefinition Height="Auto"/> 
            <RowDefinition/>
        </Grid.RowDefinitions>
        
        <controls:TabbedCommandBar>
            <controls:TabbedCommandBar.PaneFooter>
                <CommandBar Background="Transparent" DefaultLabelPosition="Right">
                    <AppBarButton Label="Share" Icon="Share"/>
                    <AppBarButton Label="Comments" Icon="Message"/>
                </CommandBar>
            </controls:TabbedCommandBar.PaneFooter>
            <controls:TabbedCommandBar.MenuItems>
                <controls:TabbedCommandBarItem Header="Home">
                    <AppBarButton Icon="Undo" Label="Undo"/>
                    <AppBarButton Icon="Redo" Label="Redo"/>
                    <AppBarButton Icon="Paste" Label="Paste"/>
                    <AppBarSeparator />
                    <AppBarElementContainer>
                        <SplitButton>
                            <PathIcon Data="F1 M 14.785156 12.089844 C 15.305989 12.089844 15.748697 12.272136 16.113281 12.636719 C 16.477863 13.001303 16.660156 13.444011 16.660156 13.964844 L 16.660156 16.464844 C 16.660156 16.972656 16.477863 17.412109 16.113281 17.783203 C 15.748697 18.154297 15.305989 18.339844 14.785156 18.339844 L 4.375 18.339844 C 3.854167 18.339844 3.411458 18.154297 3.046875 17.783203 C 2.682292 17.412109 2.5 16.972656 2.5 16.464844 L 2.5 13.964844 C 2.5 13.444011 2.682292 13.001303 3.046875 12.636719 C 3.411458 12.272136 3.854167 12.089844 4.375 12.089844 Z M 14.785156 13.339844 L 4.375 13.339844 C 4.205729 13.339844 4.059245 13.398438 3.935547 13.515625 C 3.811849 13.632812 3.75 13.782553 3.75 13.964844 L 3.75 16.464844 C 3.75 16.634115 3.811849 16.7806 3.935547 16.904297 C 4.059245 17.027994 4.205729 17.089844 4.375 17.089844 L 14.785156 17.089844 C 14.967447 17.089844 15.117187 17.027994 15.234375 16.904297 C 15.351562 16.7806 15.410155 16.634115 15.410156 16.464844 L 15.410156 13.964844 C 15.410155 13.782553 15.351562 13.632812 15.234375 13.515625 C 15.117187 13.398438 14.967447 13.339844 14.785156 13.339844 Z M 5.878906 9.980469 L 9.003906 2.070312 C 9.108072 1.822918 9.290364 1.689453 9.550781 1.669922 C 9.811197 1.650391 10 1.751303 10.117188 1.972656 L 10.175781 2.070312 L 13.28125 9.980469 C 13.346354 10.136719 13.346354 10.296225 13.28125 10.458984 C 13.216145 10.621745 13.098957 10.729167 12.929688 10.78125 C 12.786457 10.846354 12.643229 10.852865 12.5 10.800781 C 12.356771 10.748698 12.246094 10.651042 12.167969 10.507812 L 12.128906 10.429688 L 11.289062 8.339844 L 7.871094 8.339844 L 7.03125 10.429688 C 6.979167 10.585938 6.884766 10.696615 6.748047 10.761719 C 6.611328 10.826823 6.464844 10.846354 6.308594 10.820312 L 6.230469 10.78125 C 6.087239 10.729167 5.979817 10.631511 5.908203 10.488281 C 5.836588 10.345053 5.820312 10.201823 5.859375 10.058594 L 5.878906 9.980469 Z M 9.589844 3.984375 L 8.359375 7.089844 L 10.800781 7.089844 Z " VerticalAlignment="Center" HorizontalAlignment="Center" />
                            <SplitButton.Flyout>
                                <Flyout>
                                    <controls:ColorPicker Margin="-10" Color="{ThemeResource Brand-Color}"/>
                                </Flyout>
                            </SplitButton.Flyout>
                        </SplitButton>
                    </AppBarElementContainer>
                    <AppBarElementContainer>
                        <ComboBox SelectedIndex="0" MinWidth="175">
                            <ComboBoxItem Content="Arial" />
                            <ComboBoxItem Content="Calibri" />
                            <ComboBoxItem Content="JetBrains Mono" />
                            <ComboBoxItem Content="Roboto" />
                            <ComboBoxItem Content="Sergio UI" />
                            <ComboBoxItem Content="Sergio UI Semibold" />
                        </ComboBox>
                    </AppBarElementContainer>
                    <AppBarElementContainer>
                        <TextBox PlaceholderText="Size"/>
                    </AppBarElementContainer>
                    <AppBarToggleButton Icon="Bold" Label="Bold" />
                    <AppBarToggleButton Icon="Italic" Label="Italic" />
                    <AppBarToggleButton Icon="Underline" Label="Underline" />
                </controls:TabbedCommandBarItem>
                <controls:TabbedCommandBarItem Header="Insert">
                    <AppBarButton Icon="Pictures" Label="Pictures">
                        <AppBarButton.Flyout>
                            <MenuFlyout Placement="BottomEdgeAlignedLeft">
                                <MenuFlyoutItem Text="This Device">
                                    <MenuFlyoutItem.Icon>
                                        <FontIcon FontFamily="Segoe MDL2 Assets" Glyph="&#xEC4E;" />
                                    </MenuFlyoutItem.Icon>
                                </MenuFlyoutItem>
                                <MenuFlyoutItem Text="Stock Images">
                                    <MenuFlyoutItem.Icon>
                                        <FontIcon FontFamily="Segoe MDL2 Assets" Glyph="&#xE721;" />
                                    </MenuFlyoutItem.Icon>
                                </MenuFlyoutItem>
                                <MenuFlyoutItem Icon="Globe" Text="Online Pictures" />
                            </MenuFlyout>
                        </AppBarButton.Flyout>
                    </AppBarButton>
                    <AppBarButton Label="Shapes">
                        <AppBarButton.Icon>
                            <FontIcon FontFamily="Segoe UI Symbol" Glyph="&#x25A1;" />
                        </AppBarButton.Icon>
                    </AppBarButton>
                    <AppBarButton Label="Icons">
                        <AppBarButton.Icon>
                            <FontIcon FontFamily="Segoe MDL2 Assets" Glyph="&#xED58;" />
                        </AppBarButton.Icon>
                    </AppBarButton>
                    <AppBarButton Label="3D Models">
                        <AppBarButton.Icon>
                            <FontIcon FontFamily="Segoe MDL2 Assets" Glyph="&#xF158;" />
                        </AppBarButton.Icon>
                    </AppBarButton>
                    <AppBarSeparator/>
                    <AppBarButton Label="Add-ins">
                        <AppBarButton.Icon>
                            <FontIcon FontFamily="Segoe MDL2 Assets" Glyph="&#xECAA;" />
                        </AppBarButton.Icon>
                    </AppBarButton>
                    <controls:TabbedCommandBarItem.SecondaryCommands>
                        <AppBarButton Icon="Add" Label="New item" />
                    </controls:TabbedCommandBarItem.SecondaryCommands>
                </controls:TabbedCommandBarItem>
                <controls:TabbedCommandBarItem x:Name="ShapeFormat" Header="Shape Format" IsContextual="True" Visibility="Collapsed">
                    <AppBarButton Label="Shape Fill">
                        <AppBarButton.Icon>
                            <PathIcon Data="F1 M 7.5 1.464844 C 7.5 1.282553 7.561849 1.132812 7.685547 1.015625 C 7.809245 0.898438 7.955729 0.839844 8.125 0.839844 C 8.294271 0.839844 8.440755 0.898438 8.564453 1.015625 C 8.68815 1.132812 8.75 1.282553 8.75 1.464844 L 8.75 2.519531 C 8.932291 2.532553 9.114583 2.578125 9.296875 2.65625 C 9.479166 2.734375 9.641927 2.845053 9.785156 2.988281 L 12.734375 5.9375 C 13.059896 6.263021 13.222656 6.656901 13.222656 7.119141 C 13.222656 7.581381 13.059896 7.975261 12.734375 8.300781 L 9.199219 11.835938 C 8.873697 12.161459 8.479817 12.324219 8.017578 12.324219 C 7.555338 12.324219 7.161458 12.161459 6.835938 11.835938 L 3.886719 8.886719 C 3.561198 8.561198 3.398437 8.167318 3.398438 7.705078 C 3.398437 7.242839 3.561198 6.848959 3.886719 6.523438 L 7.421875 2.988281 C 7.434896 2.975261 7.447916 2.96224 7.460938 2.949219 C 7.473958 2.936199 7.486979 2.929688 7.5 2.929688 L 7.5 1.464844 Z M 7.5 5.214844 L 7.5 4.6875 L 5.097656 7.089844 L 11.972656 7.089844 C 11.959635 7.037761 11.943359 6.988933 11.923828 6.943359 C 11.904297 6.897787 11.875 6.855469 11.835938 6.816406 L 8.886719 3.886719 C 8.873697 3.860678 8.854166 3.841146 8.828125 3.828125 C 8.802083 3.815105 8.776041 3.802084 8.75 3.789062 L 8.75 5.214844 C 8.75 5.384115 8.68815 5.5306 8.564453 5.654297 C 8.440755 5.777995 8.294271 5.839844 8.125 5.839844 C 7.955729 5.839844 7.809245 5.777995 7.685547 5.654297 C 7.561849 5.5306 7.5 5.384115 7.5 5.214844 Z M 5.117188 8.339844 L 7.714844 10.957031 C 7.792969 11.035156 7.890625 11.074219 8.007812 11.074219 C 8.125 11.074219 8.222656 11.035156 8.300781 10.957031 L 10.917969 8.339844 Z M 5.488281 11.660156 L 3.75 11.660156 C 3.177083 11.660156 2.685547 11.865234 2.275391 12.275391 C 1.865234 12.685547 1.660156 13.177084 1.660156 13.75 L 1.660156 16.25 C 1.660156 16.822916 1.865234 17.314453 2.275391 17.724609 C 2.685547 18.134766 3.177083 18.339844 3.75 18.339844 L 16.25 18.339844 C 16.822916 18.339844 17.314453 18.134766 17.724609 17.724609 C 18.134766 17.314453 18.339844 16.822916 18.339844 16.25 L 18.339844 13.75 C 18.339844 13.242188 18.177082 12.799479 17.851562 12.421875 C 17.526041 12.044271 17.128906 11.803386 16.660156 11.699219 C 16.660156 11.751303 16.660156 11.800131 16.660156 11.845703 C 16.660156 11.891276 16.660156 11.933594 16.660156 11.972656 C 16.660156 12.141928 16.647135 12.307943 16.621094 12.470703 C 16.595051 12.633464 16.555988 12.792969 16.503906 12.949219 C 16.673176 13.014323 16.81315 13.11849 16.923828 13.261719 C 17.034504 13.404948 17.089844 13.567709 17.089844 13.75 L 17.089844 16.25 C 17.089844 16.484375 17.008463 16.682943 16.845703 16.845703 C 16.682941 17.008463 16.484375 17.089844 16.25 17.089844 L 3.75 17.089844 C 3.515625 17.089844 3.317057 17.008463 3.154297 16.845703 C 2.991536 16.682943 2.910156 16.484375 2.910156 16.25 L 2.910156 13.75 C 2.910156 13.515625 2.991536 13.317058 3.154297 13.154297 C 3.317057 12.991537 3.515625 12.910156 3.75 12.910156 L 6.953125 12.910156 C 6.822917 12.858073 6.699219 12.789714 6.582031 12.705078 C 6.464844 12.620443 6.354167 12.526042 6.25 12.421875 Z M 11.660156 11.972656 C 11.660156 11.777344 11.689453 11.578776 11.748047 11.376953 C 11.806641 11.175131 11.875 10.989584 11.953125 10.820312 C 12.031249 10.651042 12.11914 10.488281 12.216797 10.332031 C 12.314452 10.175781 12.415363 10.026042 12.519531 9.882812 C 12.610676 9.739584 12.705077 9.612631 12.802734 9.501953 C 12.90039 9.391276 12.98177 9.290365 13.046875 9.199219 C 13.085938 9.160156 13.13151 9.114584 13.183594 9.0625 C 13.235677 9.010417 13.274739 8.971354 13.300781 8.945312 L 13.300781 8.925781 C 13.430989 8.808594 13.580729 8.75 13.75 8.75 C 13.91927 8.75 14.06901 8.808594 14.199219 8.925781 L 14.453125 9.199219 C 14.518229 9.290365 14.599609 9.391276 14.697266 9.501953 C 14.794922 9.612631 14.889322 9.739584 14.980469 9.882812 C 15.084635 10.026042 15.185547 10.175781 15.283203 10.332031 C 15.380859 10.488281 15.46875 10.651042 15.546875 10.820312 C 15.625 10.989584 15.693359 11.175131 15.751953 11.376953 C 15.810547 11.578776 15.839844 11.777344 15.839844 11.972656 C 15.839844 12.558594 15.644531 13.069662 15.253906 13.505859 C 14.863281 13.942058 14.361979 14.160156 13.75 14.160156 C 13.13802 14.160156 12.636718 13.942058 12.246094 13.505859 C 11.855469 13.069662 11.660156 12.558594 11.660156 11.972656 Z M 13.964844 10.605469 C 13.925781 10.553386 13.889974 10.504558 13.857422 10.458984 C 13.824869 10.413412 13.789062 10.364584 13.75 10.3125 C 13.710938 10.364584 13.67513 10.413412 13.642578 10.458984 C 13.610025 10.504558 13.574218 10.553386 13.535156 10.605469 C 13.44401 10.722656 13.359375 10.846354 13.28125 10.976562 C 13.203125 11.106771 13.138021 11.230469 13.085938 11.347656 C 13.033854 11.477865 12.991535 11.595053 12.958984 11.699219 C 12.926432 11.803386 12.910156 11.894531 12.910156 11.972656 C 12.910156 12.259115 12.994791 12.486979 13.164062 12.65625 C 13.333332 12.825521 13.528645 12.910156 13.75 12.910156 C 13.971354 12.910156 14.166666 12.825521 14.335938 12.65625 C 14.505207 12.486979 14.589843 12.259115 14.589844 11.972656 C 14.589843 11.894531 14.573566 11.803386 14.541016 11.699219 C 14.508463 11.595053 14.466146 11.477865 14.414062 11.347656 C 14.361979 11.230469 14.296875 11.106771 14.21875 10.976562 C 14.140625 10.846354 14.055989 10.722656 13.964844 10.605469 Z " VerticalAlignment="Center" HorizontalAlignment="Center" />
                        </AppBarButton.Icon>
                    </AppBarButton>
                    <AppBarButton Label="Shape Outline">
                        <AppBarButton.Icon>
                            <PathIcon Data="F1 M 3.652344 11.660156 C 3.105469 11.686198 2.636719 11.901042 2.246094 12.304688 C 1.855469 12.708334 1.660156 13.190104 1.660156 13.75 L 1.660156 16.25 C 1.660156 16.822916 1.865234 17.314453 2.275391 17.724609 C 2.685547 18.134766 3.177083 18.339844 3.75 18.339844 L 16.25 18.339844 C 16.822916 18.339844 17.314453 18.134766 17.724609 17.724609 C 18.134766 17.314453 18.339844 16.822916 18.339844 16.25 L 18.339844 13.75 C 18.339844 13.177084 18.134766 12.685547 17.724609 12.275391 C 17.314453 11.865234 16.822916 11.660156 16.25 11.660156 L 11.289062 11.660156 L 10.195312 12.753906 C 10.169271 12.779948 10.139974 12.80599 10.107422 12.832031 C 10.074869 12.858073 10.045572 12.884115 10.019531 12.910156 L 16.25 12.910156 C 16.484375 12.910156 16.682941 12.991537 16.845703 13.154297 C 17.008463 13.317058 17.089844 13.515625 17.089844 13.75 L 17.089844 16.25 C 17.089844 16.484375 17.008463 16.682943 16.845703 16.845703 C 16.682941 17.008463 16.484375 17.089844 16.25 17.089844 L 3.75 17.089844 C 3.515625 17.089844 3.317057 17.008463 3.154297 16.845703 C 2.991536 16.682943 2.910156 16.484375 2.910156 16.25 L 2.910156 13.75 C 2.910156 13.59375 2.949219 13.450521 3.027344 13.320312 C 3.105469 13.190104 3.209635 13.092448 3.339844 13.027344 C 3.326823 12.923178 3.326823 12.819011 3.339844 12.714844 C 3.352865 12.610678 3.378906 12.506511 3.417969 12.402344 L 3.652344 11.660156 Z M 15.546875 2.460938 C 16.067707 2.981771 16.328125 3.613281 16.328125 4.355469 C 16.328125 5.097656 16.067707 5.729167 15.546875 6.25 L 9.609375 12.167969 C 9.492188 12.285156 9.368489 12.382812 9.238281 12.460938 C 9.108072 12.539062 8.964844 12.597656 8.808594 12.636719 L 5.234375 13.710938 C 5.013021 13.776042 4.801432 13.756511 4.599609 13.652344 C 4.397786 13.548178 4.264323 13.385417 4.199219 13.164062 C 4.173177 13.072917 4.160156 12.985026 4.160156 12.900391 C 4.160156 12.815756 4.173177 12.734375 4.199219 12.65625 L 5.332031 9.140625 C 5.384114 8.997396 5.449219 8.860678 5.527344 8.730469 C 5.605469 8.600261 5.696614 8.483073 5.800781 8.378906 L 11.738281 2.460938 C 12.272135 1.927084 12.9069 1.660156 13.642578 1.660156 C 14.378255 1.660156 15.013021 1.927084 15.546875 2.460938 Z M 12.636719 3.339844 L 6.679688 9.277344 C 6.653646 9.303386 6.627604 9.339193 6.601562 9.384766 C 6.575521 9.430339 6.549479 9.472656 6.523438 9.511719 L 5.644531 12.285156 L 8.457031 11.445312 C 8.509114 11.432292 8.557942 11.412761 8.603516 11.386719 C 8.649088 11.360678 8.684896 11.328125 8.710938 11.289062 L 14.648438 5.371094 C 14.934896 5.084636 15.078125 4.742839 15.078125 4.345703 C 15.078125 3.948568 14.934896 3.613282 14.648438 3.339844 C 14.374999 3.053386 14.039713 2.910156 13.642578 2.910156 C 13.245442 2.910156 12.910156 3.053386 12.636719 3.339844 Z " VerticalAlignment="Center" HorizontalAlignment="Center" />
                        </AppBarButton.Icon>
                    </AppBarButton>
                </controls:TabbedCommandBarItem>
            </controls:TabbedCommandBar.MenuItems>
        </controls:TabbedCommandBar>

        <Grid Grid.Row="1">
            <ToggleSwitch x:Name="ContextualToggle" IsOn="{Binding Visibility, ElementName=ShapeFormat, Converter={StaticResource BoolVisConverter}, Mode=TwoWay}"
                          OffContent="Contextual Tab Off" OnContent="Contextual Tab On"/>
        </Grid>
    </Grid>
</Page>
```

Looks like the TwoWay is required to make this work which sounds more bells (it does cause the Visibility changed event I had tried registering to fire though)... I also had to add the converter to the `XamlOnlyPage`

michael-hawker replied:
Hey @yoshiask, looks like you modified a small part of this, but not the complete workaround? I've fixed this in my local copy to match what I had above on top of your recent changes. I'll make a commit.

michael-hawker reviewed (changes requested):
I noticed there's a few todos? Are those future todos, things that should be removed, or issues that should be addressed?

Maybe best to create a new issue to track any open/remaining things you want to track long-term and then just put a `#1234` reference to the issue number in the comments with more details?

Otherwise, I think we just have to update the sample with the new binding to make it work by default... (see comments)

yoshiask commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/TabbedCommandBar/TabbedCommandBar.bind:
```diff
@@ -18,6 +23,16 @@
                     <AppBarSeparator />
+                    <AppBarElementContainer>
+                        <SplitButton>
+                            <PathIcon Data="F1 M 14.785156 12.089844 C 15.305989 12.089844 15.748697 12.272136 16.113281 12.636719 C 16.477863 13.001303 16.660156 13.444011 16.660156 13.964844 L 16.660156 16.464844 C 16.660156 16.972656 16.477863 17.412109 16.113281 17.783203 C 15.748697 18.154297 15.305989 18.339844 14.785156 18.339844 L 4.375 18.339844 C 3.854167 18.339844 3.411458 18.154297 3. ...
```
I thought I saw you working on this on stream once! I'll see how difficult it is to make a `ColorPickerAppBarRevealButtonStyle` or something. Would such a style be defined in `TabbedCommandBarItem.xaml` or `ColorPickerButton.xaml`?

michael-hawker replied:
@yoshiak I'd just keep it in the `TabbedCommandBarItem.xaml` with the rest of them. Then there's a central place for those that are specific to the TabbedCommandBar. It's [mostly just a default DropDownButton style](https://github.com/windows-toolkit/WindowsCommunityToolkit/blob/master/Microsoft.Toolkit.Uwp.UI.Controls/ColorPicker/ColorPickerButton.xaml), though I forget what I had to tweak, I think I had to muck with it to support the transparency background or something.

yoshiask commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/TabbedCommandBar/TabbedCommandBar.bind:
```diff
@@ -78,7 +93,25 @@
+                                               Visibility="{Binding ElementName=ContextualToggle, Path=IsOn}">
+                    <AppBarButton Label="Shape Fill">
+                        <AppBarButton.Icon>
+                            <PathIcon Data="F1 M 7.5 1.464844 C 7.5 1.282553 7.561849 1.132812 7.685547 1.015625 C 7.809245 0.898438 7.955729 0.839844 8.125 0.839844 C 8.294271 0.839844 8.440755 0.898438 8.564453 1.015625 C 8.68815 1.132812 8.75 1.282553 8.75 1.464844 L 8.75 2.519531 C 8.932291 2.532553 9.114583 2.578125 9.296875 2.65625 C 9.479166 2.734375 9.641927 2.845053 9.785156 2.988281 L 1 ...
```
I discovered a Segoe MDL2 Assets glyph for text color, but not shape fill or outline. I'd actually prefer to just use a different action there that we do have an icon for, like maybe "Crop" and "Rotate" from the "Picture Format" tab.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/TabbedCommandBar/TabbedCommandBar.cs:
```diff
@@ -0,0 +1,102 @@
+            _previousSelectedItem.RegisterPropertyChangedCallback(TabbedCommandBarItem.VisibilityProperty, SelectedItemVisibilityChanged);
+
+            // Set the ribbon background and start the transition animation
+            //_ribbonContentBorder.Background = item.Background;
```
```suggestion
```
Remove commented line causing StyleCop Issue.

michael-hawker reviewed (approved):
Alright, think we're good. @Kyaa-dost want to take it for a final spin with the fix for the Sample in the morning and we can get it merged?

@RosarioPulella after that mind updating the `dev/split-controls` branch to move this to `Core`?

yoshiask commented:
@michael-hawker I merged master back in and the bot removed the `zap` tag
